File: 3.- Codificación/frontend/src/api/Libros.js
import axios from 'axios';

const baseURL = 'http://localhost:8000'; // La URL base de tu servidor FastAPI

// Clave de cifrado generada en el servidor
const key = '6A2VDZYE6tswbrP476M89Qoktl2Is9wTXuzIB-dCntY=';

const decryptData = (encryptedData) => {
    // Crea una instancia de Fernet utilizando la clave
    const cipher = new Fernet(key);
    try {
        // Desencripta los datos
        const decryptedData = cipher.decrypt(encryptedData);
        // Devuelve los datos desencriptados como una cadena UTF-8
        return decryptedData.toString('utf-8');
    } catch (error) {
        console.error('Error al desencriptar los datos:', error);
        return null; // Devuelve null si hay un error en la desencriptación
    }
};

// Función para obtener todos los libros
export const obtenerLibros = async () => {
    try {
        const response = await axios.get(`${baseURL}/libros/`);
        console.log(response.data);
        return response.data.data;
    } catch (error) {
        console.error('Error al obtener libros:', error);
        throw error;
    }
};

// Función para crear un libro
export const crearLibro = async (libro) => {
    try {
        const response = await axios.post(`${baseURL}/libro/`, libro);
        return response.data;
    } catch (error) {
        console.error('Error al crear libro:', error);
        throw error;
    }
};

// Función para obtener un libro por su ID
export const obtenerLibroPorId = async (libroId) => {
    try {
        const response = await axios.get(`${baseURL}/libros/${libroId}`);
        return response.data.data;
    } catch (error) {
        console.error('Error al obtener libro por ID:', error);
        throw error;
    }
};

// Función para obtener libros por estado
export const obtenerLibrosPorEstado = async (estado) => {
    try {
        const response = await axios.get(`${baseURL}/libros/estado/${estado}`);
        console.log(response.data.data);
        return response.data.data;
    } catch (error) {
        console.error('Error al obtener libros por estado:', error);
        throw error;
    }
};

// Función para actualizar un libro
export const actualizarLibro = async (libroId, libro) => {
    try {
        const response = await axios.put(`${baseURL}/libros/${libroId}`, libro);
        return response.data;
    } catch (error) {
        console.error('Error al actualizar libro:', error);
        throw error;
    }
};

// Función para actualizar el estado de un libro
export const actualizarEstadoLibro = async (libroId, estado) => {
    try {
        const response = await axios.put(`${baseURL}/libros/${libroId}/estado?estado=${estado}`);
        return response.data;
    } catch (error) {
        console.error('Error al actualizar estado del libro:', error);
        throw error;
    }
};

// Función para eliminar un libro
export const eliminarLibro = async (libroId) => {
    try {
        const response = await axios.delete(`${baseURL}/libros/${libroId}`);
        return response.data;
    } catch (error) {
        console.error('Error al eliminar libro:', error);
        throw error;
    }
};

// Función para obtener libros por título
export const obtenerLibrosPorTitulo = async (titulo) => {
    try {
        const response = await axios.get(`${baseURL}/libros/titulo/?titulo=${titulo}`);
        return response.data;
    } catch (error) {
        console.error('Error al obtener libros por título:', error);
        throw error;
    }
};

// Función para obtener libros por autor
export const obtenerLibrosPorAutor = async (autor) => {
    try {
        const response = await axios.get(`${baseURL}/libros/autor/?autor=${autor}`);
        return response.data;
    } catch (error) {
        console.error('Error al obtener libros por autor:', error);
        throw error;
    }
};

File: 3.- Codificación/microservicio-libros/apilibros/routers/libros.py
from fastapi import APIRouter, HTTPException, Query
from models import Libro
from database import db_cursor, db_connection
import json
from Crypto.Cipher import AES

key = b'mi_clave_secreta'

import binascii

def encrypt(key, data):
    cipher = AES.new(key, AES.MODE_EAX)
    ciphertext, tag = cipher.encrypt_and_digest(data.encode())

    encrypted_data = cipher.nonce + tag + ciphertext
    return binascii.hexlify(encrypted_data)

def decrypt(key, encrypted_data):
    encrypted_bytes = binascii.unhexlify(encrypted_data)
    nonce = encrypted_bytes[:AES.block_size]
    tag = encrypted_bytes[AES.block_size:AES.block_size * 2]
    ciphertext = encrypted_bytes[AES.block_size * 2:]

    cipher = AES.new(key, AES.MODE_EAX, nonce)
    plaintext = cipher.decrypt_and_verify(ciphertext, tag)

    return plaintext


router = APIRouter()

# CRUD de libros
@router.post("/libro/")
async def crear_libro(libro: Libro):
    query = "INSERT INTO libros (titulo, ISBN, año_publicacion, editorial, autor, estado) VALUES (%s, %s, %s, %s, %s, %s)"
    values = (libro.titulo, libro.ISBN, libro.año_publicacion, libro.editorial, libro.autor, libro.estado)
    db_cursor.execute(query, values)
    db_connection.commit()
    return {"mensaje": "Libro creado exitosamente", "libro_id": db_cursor.lastrowid}

@router.get("/libros/")
async def obtener_libros():
    query = "SELECT * FROM libros"
    db_cursor.execute(query)
    libros = db_cursor.fetchall()
    lista_libros = []
    for libro in libros:
        libro_dict = {
            "libro_id": str(libro[0]),
            "titulo": libro[1],
            "ISBN": libro[2],
            "año_publicacion": str(libro[3]),
            "editorial": libro[4],
            "autor": libro[5],
            "estado": libro[6]
        }
        encrypted_libro = encrypt(key, json.dumps(libro_dict))
        lista_libros.append(encrypted_libro)
    return {"data": lista_libros}

# OBTENER LIBRO POR ID
@router.get("/libros/{libro_id}")
async def obtener_libro(libro_id: int):
    query = "SELECT * FROM libros WHERE libro_id = %s"
    db_cursor.execute(query, (libro_id,))
    libro = db_cursor.fetchone()
    if not libro:
        raise HTTPException(status_code=404, detail="Libro no encontrado")
    libro_dict = {
        "libro_id": str(libro[0]),
        "titulo": libro[1],
        "ISBN": libro[2],
        "año_publicacion": str(libro[3]),
        "editorial": libro[4],
        "autor": libro[5],
        "estado": libro[6]
    }
    encrypted_libro = encrypt(key, json.dumps(libro_dict))
    return {"data": encrypted_libro}

# OBTENER LIBRO POR ESTADO
@router.get("/libros/estado/{estado}")
async def obtener_libros_por_estado(estado: str):
    query = "SELECT * FROM libros WHERE estado = %s"
    db_cursor.execute(query, (estado,))
    libros = db_cursor.fetchall()
    
    lista_libros = []
    for libro in libros:
        libro_dict = {
            "libro_id": str(libro[0]),
            "titulo": libro[1],
            "ISBN": libro[2], 
            "año_publicacion": str(libro[3]),
            "editorial": libro[4],
            "autor": libro[5],
            "estado": libro[6]
        }
        encrypted_libro = encrypt(key, json.dumps(libro_dict))
        lista_libros.append(encrypted_libro)
        
    return {"data": lista_libros}

# ACTUALIZAR LIBROS
@router.put("/libros/{libro_id}")
async def actualizar_libro(libro_id: int, libro: Libro):
    query = "UPDATE libros SET titulo = %s, ISBN = %s, año_publicacion = %s, editorial = %s, autor = %s, estado = %s WHERE libro_id = %s"
    values = (libro.titulo, libro.ISBN, libro.año_publicacion, libro.editorial, libro.autor, libro.estado, libro_id)
    db_cursor.execute(query, values)
    db_connection.commit()
    return {"mensaje": "Libro actualizado exitosamente"}

# ACTUALIZAR ESTADO DE LOS LIBROS
@router.put("/libros/{libro_id}/estado")
async def actualizar_estado_libro(libro_id: int, estado: str):
    query = "UPDATE libros SET estado = %s WHERE libro_id = %s"
    values = (estado, libro_id)
    db_cursor.execute(query, values)
    db_connection.commit()
    return {"mensaje": "Estado del libro actualizado exitosamente"}

# ELIMINAR LIBROS
@router.delete("/libros/{libro_id}")
async def eliminar_libro(libro_id: int):
    query = "DELETE FROM libros WHERE libro_id = %s"
    db_cursor.execute(query, (libro_id,))
    db_connection.commit()
    return {"mensaje": "Libro eliminado exitosamente"}

# RUTA PARA OBTENER LIBROS POR TÍTULO
@router.get("/libros/titulo/")
async def obtener_libros_por_titulo(titulo: str = Query(..., min_length=1)):
    query = "SELECT * FROM libros WHERE titulo LIKE %s"
    db_cursor.execute(query, ('%' + titulo + '%',))
    libros = db_cursor.fetchall()
    
    lista_libros = []
    for libro in libros:
        libro_dict = {
            "libro_id": str(libro[0]),
            "titulo": libro[1],
            "ISBN": libro[2], 
            "año_publicacion": str(libro[3]),
            "editorial": libro[4],
            "autor": libro[5],
            "estado": libro[6]
        }
        encrypted_libro = encrypt(key, json.dumps(libro_dict))
        lista_libros.append(encrypted_libro)
        
    return {"data": lista_libros}

# RUTA PARA OBTENER LIBROS POR AUTOR
@router.get("/libros/autor/")
async def obtener_libros_por_autor(autor: str = Query(..., min_length=1)):
    query = "SELECT * FROM libros WHERE autor LIKE %s"
    db_cursor.execute(query, ('%' + autor + '%',))
    libros = db_cursor.fetchall()
    
    lista_libros = []
    for libro in libros:
        libro_dict = {
            "libro_id": str(libro[0]),
            "titulo": libro[1],
            "ISBN": libro[2], 
            "año_publicacion": str(libro[3]),
            "editorial": libro[4],
            "autor": libro[5],
            "estado": libro[6]
        }
        encrypted_libro = encrypt(key, json.dumps(libro_dict))
        lista_libros.append(encrypted_libro)
        
    return {"data": lista_libros}